metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java

```diff
@@ -38,27 +40,35 @@ public EnricherImpl(EntityMergeEngine entityMergeEngine,
   public void enrichment(RDF rdf) throws EnrichmentException {
     // Extract values and references from the RDF for enrichment
     LOGGER.debug("Extracting values and references from RDF for enrichment...");
-    final List<InputValue> valuesForEnrichment = extractValuesForEnrichment(rdf);
+    final List<Pair<SearchValue, EnrichmentFields>> valuesForEnrichment = extractValuesForEnrichment(
+        rdf);
     final Map<String, Set<EnrichmentFields>> referencesForEnrichment = extractReferencesForEnrichment(
         rdf);
 
     // Get the information with which to enrich the RDF using the extracted values and references
     LOGGER.debug("Using extracted values and references to gather enrichment information...");
-    final EnrichmentResultList valueEnrichmentInformation = enrichValues(valuesForEnrichment);
-    final List<EnrichmentBaseWrapper> referenceEnrichmentInformation = enrichReferences(
+    final List<Pair<EnrichmentResultBaseWrapper, EnrichmentFields>> valueEnrichmentInformation = enrichValues(
+        valuesForEnrichment);
+    final List<EnrichmentResultBaseWrapper> referenceEnrichmentInformation = enrichReferences(
         referencesForEnrichment.keySet());
 
     // Merge the acquired information into the RDF
     LOGGER.debug("Merging Enrichment Information...");
-    if (valueEnrichmentInformation != null && CollectionUtils
-        .isNotEmpty(valueEnrichmentInformation.getEnrichmentBaseWrapperList())) {
-      entityMergeEngine
-          .mergeEntities(rdf, valueEnrichmentInformation.getEnrichmentBaseWrapperList());
-    }
-    if (!referenceEnrichmentInformation.isEmpty()) {
-      final List<EnrichmentBase> entities = referenceEnrichmentInformation.stream()
-          .map(EnrichmentBaseWrapper::getEnrichmentBase).collect(Collectors.toList());
-      entityMergeEngine.mergeEntities(rdf, entities, referencesForEnrichment);
+    if (valueEnrichmentInformation != null) {
+      final Set<EnrichmentFields> enrichmentFieldsSet = valueEnrichmentInformation.stream()
+          .map(Pair::getSecond).collect(Collectors.toSet());
+      final List<EnrichmentResultBaseWrapper> enrichmentResultBaseWrapperList = valueEnrichmentInformation
+          .stream().map(Pair::getFirst).collect(Collectors.toList());
+
+      if (CollectionUtils.isNotEmpty(enrichmentResultBaseWrapperList)) {
+        entityMergeEngine
+            .mergeEntities(rdf, enrichmentResultBaseWrapperList, enrichmentFieldsSet);
```

Comment:
I think there is a problem here. By calling it like this, you add all results and all fields, and you lose the mapping. Instead you have to loop through the pairs and call the mergeEntity method multiple times to make sure that results are only added to the fields that they belong to.

Reply:
Should I change the signature of mergeEntities then? To Accept EnrichmentBase and EnrichmentFields alone? Or should I be careful with that inside mergeEntities?

Reply:
No, I think you should be careful with that here in this method. 

I think this should be wrapped in a loop over all pairs in `valueEnrichmentInformation` and it should be called once for each pair. Do you agree?

Reply:
It is done